src/Aspire.Hosting.AWS.Provisioning/CloudFormation/CloudFormationLifeCycle.cs

```diff
@@ -0,0 +1,237 @@
+            {
+                mostRecentEventId = events[0].EventId;
+            }
+
+            for (int i = events.Count - 1; i >= 0; i--)
+            {
+                string line =
+                    events[i].Timestamp.ToString("g", CultureInfo.InvariantCulture).PadRight(TIMESTAMP_WIDTH) + " " +
+                    events[i].LogicalResourceId.PadRight(LOGICAL_RESOURCE_WIDTH) + " " +
+                    events[i].ResourceStatus.ToString(CultureInfo.InvariantCulture).PadRight(RESOURCE_STATUS);
+
+                if (!events[i].ResourceStatus.ToString(CultureInfo.InvariantCulture).EndsWith(IN_PROGRESS_SUFFIX) && !string.IsNullOrEmpty(events[i].ResourceStatusReason))
+                {
+                    line += " " + events[i].ResourceStatusReason;
+                }
+
+                if(minTimeStampForEvents < events[i].Timestamp)
+                {
+                    minTimeStampForEvents = events[i].Timestamp;
+                }
+
+                logger.LogInformation(line);
+            }
+
+        } while (stack.StackStatus.ToString(CultureInfo.InvariantCulture).EndsWith(IN_PROGRESS_SUFFIX));
+
+        return stack;
+    }
+
+    private static async Task<List<StackEvent>> GetLatestEventsAsync(IAmazonCloudFormation cfClient, string stackName, DateTime minTimeStampForEvents, string mostRecentEventId, CancellationToken cancellationToken)
+    {
+        var noNewEvents = false;
+        var events = new List<StackEvent>();
+        DescribeStackEventsResponse? response = null;
+        do
+        {
+            var request = new DescribeStackEventsRequest() { StackName = stackName };
+            if (response != null)
+            {
+                request.NextToken = response.NextToken;
+            }
+
+            try
+            {
+                response = await cfClient.DescribeStackEventsAsync(request, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                throw new AWSProvisioningException($"Error getting events for CloudFormation stack: {e.Message}", e);
+            }
+            foreach (var evnt in response.StackEvents)
+            {
+                if (string.Equals(evnt.EventId, mostRecentEventId) || evnt.Timestamp < minTimeStampForEvents)
+                {
+                    noNewEvents = true;
+                    break;
+                }
+
+                events.Add(evnt);
+            }
+
+        } while (!noNewEvents && !string.IsNullOrEmpty(response.NextToken));
+
+        return events;
+    }
+
+    private static string ComputeSHA256(string templateBody)
+    {
+        using var sha256Hash = SHA256.Create();
+
+        var bytes = sha256Hash.ComputeHash(Encoding.UTF8.GetBytes(templateBody));
+        var builder = new StringBuilder();
+        for (int i = 0; i < bytes.Length; i++)
+        {
+            builder.Append(bytes[i].ToString("x2", CultureInfo.InvariantCulture));
+        }
+        return builder.ToString();
```

Comment:
If lower vs. upper isn't important, this could just be:

```suggestion
        return Convert.ToHexString(bytes);
```

Reply:
Done by added `.ToLower(CultureInfo.InvariantCulture)`. Just feel like a SHA should be lowercase, not sure why I feel that way.

Reply:
https://github.com/dotnet/runtime/issues/60393